---
title: "Myc;Ptenfl TNBC mouse model bulk RNA-seq analysis & PCA & mIHC"
author: "Zinab O. Doha" and "Nick Calistri"
date: "01/28/2021"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: true
    highlight: tango
---

# Set up

## Load Libraries

```{r}
library(tidyverse)
library(dplyr)
library(DESeq2)
```

## Set up color coding

```{r}
# Set up metadata colors
col_dict <- tibble(names = c('SP', 'SR'),
                   colors = c('#F8766D', '#619CFF'))

# Set up range to use for expression heatmaps + color sequence
exp_breaks <- seq(from = -3, to = 3, by = 0.05)
exp_colors <- colorRampPalette(c('purple', 'black', 'yellow'))(length(exp_breaks))

```



## Import and format data

```{r}
# Load sample metadata
sample_hist_dict <- read_csv('bulkrnaseq_sample_hist_dict.csv') %>%
  mutate(sample = str_remove(sample, pattern = '_TPM'))

# Load TPM gene expression matrix
tpm <- read.csv("data/expression_file.csv")
counts <- read.delim('data/NeuNTandPTEN_CountsByGeneSymbol_Matrix.txt')

# format TPM gene expresison matrix as dataframe with unique row names
rownames(counts) <- counts$GeneSymbol ## row names should be gene names

counts <- counts %>%
  dplyr::select(-GeneSymbol)

counts_mat <- data.matrix(counts)

# Filter to remove rows (genes) with no expression
counts_mat <- counts_mat[rowSums(counts_mat)!=0, ]

# Number of filtered genes
print(paste('Number of filtered (0 counts) genes:', nrow(counts) - nrow(counts_mat)))

```


# MycPten;fl Analysis  

## Convert to DESeq2 format

```{r}
# Format metadata
coldata <- data.frame(row.names = sample_hist_dict$sample,
                      phenotype = sample_hist_dict$histology_phenotype)

# Filter counts matrix to only include samples of interest
counts_filt <- counts_mat[,row.names(coldata)]

# Validate coldata aligns with counts column names
all(colnames(counts_filt) == rownames(coldata))

# Create deseq object
dds <- DESeqDataSetFromMatrix(countData = counts_filt,
                              colData = coldata,
                              design = ~phenotype)

# Filter to only genes that have >10 counts in 5 samples
keep <- rowSums(counts(dds) >= 10) >= 5
dds <- dds[keep,]

# Assign Stromal Rich as reference
dds$phenotype <- relevel(dds$phenotype, ref = 'SR')
```



## Identify variable genes & Compute PCA

```{r}

vsd <- vst(dds, blind = TRUE)
plotPCA(vsd, intgroup = 'phenotype', ntop = 1000)

# Manual PCA

## Select ntop most variable genes
ntop <- 1000
rv <- genefilter::rowVars(assay(vsd))
topvar_feats <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]

## Perform PCA
pca <- prcomp(t(assay(vsd)[topvar_feats,]), center = TRUE, scale = FALSE)

## Conver to tibble and visualize
pca_tibble <- as_tibble(pca$x) %>%
  mutate(sample = row.names(pca$x)) %>%
  mutate(phenotype = plyr::mapvalues(x = sample,
                                     from = sample_hist_dict$sample,
                                     to = sample_hist_dict$histology_phenotype))

ggplot(pca_tibble, aes(x = PC1*summary(pca)$importance[2,1], y = PC2*summary(pca)$importance[2,2], color = phenotype))+
  geom_point()+
  theme_bw()+
  ggrepel::geom_text_repel(aes(label = str_split(sample, pattern = '_', simplify = TRUE)[,5]))+
  ggtitle('Principal Component Analysis \n (Axes scaled by proportion of variance)')+
  xlab(paste0('PC1 \n ', round(summary(pca)$importance[2,1]*100,0), '% of variance'))+
  ylab(paste0('PC2 \n ', round(summary(pca)$importance[2,2]*100,0), '% of variance'))+
  coord_fixed()+
  theme(plot.title = element_text(hjust = 0.5))+
  scale_color_manual(values = col_dict$colors)

```

Caption: Principal component analysis showing the first two principal components computed on the top 1000 variable genes.


## Perform DESeq2 deg 

```{r}
dds <- DESeq(dds)
```

## Visualize DEGs

```{r}
res <- results(dds)

resLFC <- lfcShrink(dds, coef = 'phenotype_SP_vs_SR', type = 'apeglm')

# Convert to tibble and plot
res_tibble <- results(dds) %>%
  as_tibble() %>%
  mutate(gene = rownames(results(dds)))

ggplot(res_tibble, aes(x = baseMean, y = log2FoldChange, color = padj < 0.05))+
  geom_point()+
  scale_x_log10()

plotMA(results(dds), ylim = c(-5, 5))

plotMA(resLFC)
```


## Variable feature heatmap

```{r, fig.asp = 1}
# Format column/row annotation
col_anno <- data.frame(row.names = sample_hist_dict$sample,
                       subtype = sample_hist_dict$histology_phenotype)

row_anno <- data.frame(row.names = row.names(assay(vsd)[topvar_feats,]))

row_anno$pc1_weight <- ifelse(pca$rotation[,1] > 0, 'positive', 'negative')

# Format color annotations
pc_weight_cols <- c('pink', 'cyan')
names(pc_weight_cols) <- c('positive', 'negative')

subtype_cols <- col_dict$colors
names(subtype_cols) <- c('SP', 'SR')


anno_colors <- list(pc1_weight = pc_weight_cols,
                    subtype = subtype_cols)

# Generate heatmap
pheatmap::pheatmap(assay(vsd)[topvar_feats,],
                   scale = 'row',
                   annotation_col = col_anno,
                   annotation_row = row_anno,
                   show_rownames = FALSE,
                   annotation_colors = anno_colors,
                   main = 'Gene Expression \n (Top 1000 variable genes)',
                   labels_col = str_split(colnames(assay(vsd)),
                                          pattern = '_',
                                          simplify = TRUE)[,5],
                   color = exp_colors,
                   breaks = exp_breaks)

```

Caption: Heatmap showing the top 1000 variable genes used in principal component analysis. Counts were normalized with a Variance Stabilizing Transformation, and then z scored across samples. Row annotation shows whether each individual gene had a positive or negative weight in PC1 (~60% of total PCA variance).

## Sample-sample correlation

```{r, fig.asp = 1}
sampleCors <- cor(assay(vsd)[topvar_feats,])

pheatmap::pheatmap(sampleCors,
                   annotation_row = col_anno,
                   annotation_col = col_anno,
                   display_numbers = round(sampleCors,2),
                   main = 'Sample-sample Pearson correlation \n Top 1000 variable features \n VST transformed counts',
                   labels_col = str_split(colnames(sampleCors),
                                          pattern = '_',
                                          simplify = TRUE)[,5],
                   labels_row = str_split(rownames(sampleCors),
                                          pattern = '_',
                                          simplify = TRUE)[,5],
                   annotation_colors = anno_colors)
```



# TNBC subtype - MycPten;fl correlation

## Load data

```{r}
centroids <- read.delim("data/signature2.txt", row.names = 2)[,-1]
```

## Map Human genes to murine

```{r}
library(biomaRt)

human <- useEnsembl('ensembl', dataset = 'hsapiens_gene_ensembl', host = 'https://dec2021.archive.ensembl.org')
mouse <- useEnsembl('ensembl', dataset = 'mmusculus_gene_ensembl', host = 'https://dec2021.archive.ensembl.org')

hs_mm <- getLDS(attributes = c("external_gene_name"),
       mart = human,
       attributesL = c("external_gene_name"),
       martL = mouse)

# Create centroids_mm replacing human gene names with murine
centroids_mm <- centroids

rownames(centroids_mm) <- plyr::mapvalues(x = rownames(centroids),
                from = hs_mm$Gene.name,
                to = hs_mm$Gene.name.1,
                warn_missing = FALSE)
```


## Compute correlation

640x300 px
```{r}
# Find shared homologs mapped in mouse bulk RNA-seq
shared_homologs <- rownames(centroids_mm)[rownames(centroids_mm) %in% rownames(assay(vsd))]

# Scale values and subset genes to shared homologs where z score is not zero
vsd_z <- t(scale(t(assay(vsd)), center=TRUE, scale=TRUE))

valid_genes <- rownames(vsd_z[shared_homologs,])[! is.na(rowSums(vsd_z[shared_homologs,]))]

vsd_z <- vsd_z[valid_genes,]
centroids_mm <- t(scale(t(centroids_mm[valid_genes,])))

# Run correletion
mp_cor <- cor(vsd_z, centroids_mm, method="spearman")
mp_cor_melt <- reshape2::melt(as.matrix(mp_cor))
colnames(mp_cor_melt) <- c("Sample", "Subtype", "Spearman")

# Bar plot of results
ggplot(mp_cor_melt,aes(x=Sample, y=Spearman, fill=Subtype)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) +
  guides(fill=guide_legend(ncol=1))

# Heatmap of results

pheatmap::pheatmap(t(mp_cor),
                   annotation_col = col_anno,
                   annotation_colors = anno_colors,
                   main = 'Spearman correlation with TNBC subtype centroids',
                   labels_col = str_split(rownames(mp_cor),
                                          pattern = '_',
                                          simplify = TRUE)[,5],
                   cellwidth = 20,
                   cellheight = 20)

```

# Chemokine Heatmap

640x360 px

```{r, fig.asp = 1}
# Format column/row annotation
col_anno <- data.frame(row.names = sample_hist_dict$sample,
                       subtype = sample_hist_dict$histology_phenotype)



# Format color annotations

subtype_cols <- col_dict$colors
names(subtype_cols) <- c('SP', 'SR')

anno_colors <- list(subtype = subtype_cols)

# Select cytokines of interest
coi <- c('Ccl5', 'Ccl7', 'Ccl8', 'Ccl17', 'Ccl20', 'Ccl22', 'Ccl11')

# Plot heatmap
pheatmap::pheatmap(assay(vsd)[coi[coi %in% row.names(vsd)],],
                   scale = 'row',
                   annotation_col = col_anno,
                   annotation_colors = anno_colors,
                   labels_col = str_split(colnames(assay(vsd)),
                                          pattern = '_',
                                          simplify = TRUE)[,5],
                   color = exp_colors,
                   breaks = exp_breaks,
                   main = 'Prognostic chemokine expression',
                   cellwidth = 20,
                   cellheight = 20)

# display DEG results for these genes

resLFC[coi,] %>%
  as_tibble(., rownames = 'gene') %>%
  dplyr::filter(padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  knitr::kable()


```

# Cell state heatmap

720x1080 px

```{r, fig.asp = 2}
cellstate_markers <- read_csv('epithelial_markers_risom_2020_tjoci.csv') %>%
  mutate(mm_gene = plyr::mapvalues(x = gene,
                                   from = hs_mm$Gene.name,
                                   to = hs_mm$Gene.name.1,
                                   warn_missing = FALSE))

cs_filt <- cellstate_markers %>%
  filter(mm_gene %in% row.names(vsd))

row_anno <- data.frame(row.names = cs_filt$mm_gene,
                       associated_state = cs_filt$association)

# Plot heatmap
pheatmap::pheatmap(assay(vsd)[cs_filt$mm_gene,],
                   scale = 'row',
                   annotation_col = col_anno,
                   annotation_row = row_anno,
                   annotation_colors = anno_colors,
                   cluster_rows = FALSE,
                   labels_col = str_split(colnames(assay(vsd)),
                                          pattern = '_',
                                          simplify = TRUE)[,5],
                   color = exp_colors,
                   breaks = exp_breaks,
                   main = 'Epithelial state markers',
                   gaps_row = c(14,24),
                   cellheight = 15,
                   cellwidth = 15)

# display significant DEG results for these genes

resLFC[cs_filt$mm_gene,] %>%
  as_tibble(., rownames = 'gene') %>%
  dplyr::filter(padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  knitr::kable()
```


# GSEA

## Load libraries
```{r}
library(clusterProfiler)
library(org.Mm.eg.db)
library(msigdbr)
```

## Select mouse msigdb hallmark gene sets

```{r}
h_mm <- msigdbr(species = 'mouse', category = 'H')

h_mm_t2g <- h_mm %>%
  dplyr::select(gs_name, gene_symbol)
```

## Prepare ranked gene_list from resLFC
```{r}
deg <- as_tibble(resLFC) %>%
  filter(! is.na(log2FoldChange)) %>%
  mutate(gene = row.names(resLFC)) %>%
  arrange(desc(log2FoldChange))

gene_list <- deg$log2FoldChange
names(gene_list) <- deg$gene
```


## Perform GSEA on msigdb hallmark

```{r}
em <- GSEA(gene_list,
           TERM2GENE = h_mm_t2g,
           pvalueCutoff = 1,
           eps = 0,
           nPermSimple = 10000)

em_tibble <- em@result

em_tibble %>%
  dplyr::select(NES, p.adjust, pvalue) %>%
  arrange(NES) %>%
  DT::datatable()
```

## Visualize top 3 Stromal Poor pathways

720x1080 px (each)

```{r, fig.asp = 1}
sp_top3 <- em@result %>%
  filter(p.adjust < 0.05) %>%
  filter(NES > 0) %>%
  arrange(desc(NES)) %>%
  slice_head(n = 3)

for(i in sp_top3$ID){
  
  curr_hm <- as_tibble(em) %>%
    filter(ID == i)
  
  gseaplot(em, by = 'all',
           title = paste0(curr_hm$Description,
                          '\n NES: ',
                          round(curr_hm$NES,2),
                          '\n P.adj: ',
                          signif(curr_hm$p.adjust,2)),
           geneSetID = i) %>%
    print()
}

sr_top3 <- em@result %>%
  filter(p.adjust < 0.05) %>%
  filter(NES < 0) %>%
  arrange(NES) %>%
  slice_head(n = 3)

for(i in sr_top3$ID){
  
  curr_hm <- as_tibble(em) %>%
    filter(ID == i)
  
  gseaplot(em, by = 'all',
           title = paste0(curr_hm$Description,
                          '\n NES: ',
                          round(curr_hm$NES,2),
                          '\n P.adj: ',
                          signif(curr_hm$p.adjust,2)),
           geneSetID = i) %>%
    print()
}
```

## GSEA barplot

1080x720 px
```{r}
sig_hm <- em@result %>%
  filter(p.adjust < 0.05) %>%
  arrange(NES)

sig_hm <- sig_hm %>%
  mutate(ID = factor(ID, levels = sig_hm$ID)) %>%
  mutate(association = ifelse(NES > 0, 'SP', 'SR'))

ggplot(sig_hm, aes(x = NES, y = ID, fill = association))+
  geom_col()+
  theme_bw()+
  scale_color_manual(values = col_dict$colors)+
  ggtitle('MSigDB enriched Hallmarks')+
  theme(plot.title = element_text(hjust = 0.5, size = 20))
```

# Save output

```{r}
# DEGs
resLFC_tibble <- as_tibble(resLFC,
                           rownames = 'gene')

write_csv(x = resLFC_tibble,
          file = 'bulk_mycptenfl_degs.csv')

# GSEA results

write_csv(x = em@result,
          file = 'bulk_mycptenfl_msigdb_gsea.csv')

```

# mIHC analysis

```{r}
library("gplots")
library("RColorBrewer") 
library("pheatmap")

core_h <- read.csv("core.csv",row.names = 1)


inputc <- as.matrix(t(core_h))

col_anno <- data.frame(histology = str_split(colnames(inputc), pattern = '_', simplify = TRUE)[,1], row.names = colnames(inputc))


pheatmap(inputc,scale="row", annotation_col = col_anno, main = "pheatmap row annotation")
```

##Row scaling Core
```{r}
histology        <- c("#FC717F", "#00A5FF")
names(histology) <- c("SP", "SR")
anno_colors <- list(histology = histology)
pheatmap(inputc, scale="row", annotation = col_anno, annotation_colors = anno_colors, fontsize_row = 12, main = 'Core_row scaled')
```

## Row scaling_Border

```{r}
border <- read.csv("border.csv",row.names = 1)


inputb <- as.matrix(t(border))

col_anno <- data.frame(histology = str_split(colnames(inputb), pattern = '_', simplify = TRUE)[,1], row.names = colnames(inputb))

histology        <- c("#FC717F", "#00A5FF")
names(histology) <- c("SP", "SR")
anno_colors <- list(histology = histology)
dcols = dist(t(inputb), method = "minkowski")

pheatmap(inputb, scale="row", annotation = col_anno, annotation_colors = anno_colors, fontsize_row = 12, clustering_distance_cols = dcols, main = 'Border_row scaled' , legend = TRUE)
```

## Row scaling preiphry
```{r}
perph <- read.csv("perph.csv",row.names = 1)


inputp <- as.matrix(t(perph))

col_anno <- data.frame(histology = str_split(colnames(inputp), pattern = '_', simplify = TRUE)[,1], row.names = colnames(inputp))

histology        <- c("#FC717F", "#00A5FF")
names(histology) <- c("SP", "SR")
anno_colors <- list(histology = histology)
pheatmap(inputp, scale="row", annotation = col_anno, annotation_colors = anno_colors, fontsize_row = 12, main = 'Periphery_row scaled' )
```


# Metastisis Rate
```{r}
library(ggplot2)
meta <- read.csv("meta.csv")
my_ggp <- ggplot (data = meta,
       aes (x = Tumor,
            y = Metastisis_rate, fill = Tumor)) +
           geom_bar(stat = "Identity") +
  ggtitle("Metastisis Rate")
my_ggp 

my_ggp + theme(text = element_text(size = 18
                                  )) + geom_violin(colour = "yellow", alpha = 0.1)
```

# sessionInfo

```{r}
sessionInfo()
```

